rust/psibase_macros/psibase-macros-lib/src/service_macro/actions.rs

```diff
@@ -190,3 +192,114 @@ pub fn process_action_schema(
         actions.insert(#name_str.to_string(), #psibase_mod::fracpack::FunctionType{ params: builder.insert::<#action_mod::#name>(), result: #ret });
     }
 }
+
+pub struct PreAction {
+    pub fn_name: Option<Ident>,
+    pub exclude: PathList,
+}
+
+impl PreAction {
+    pub fn has_pre_action(&self) -> bool {
+        self.fn_name.is_some()
+    }
+}
+
+impl Default for PreAction {
+    fn default() -> Self {
+        PreAction {
+            fn_name: None,
+            exclude: PathList::from(vec![]),
+        }
+    }
+}
+
+fn is_pre_action_attr(attr: &Attribute) -> bool {
+    AttrStyle::Outer == attr.style && attr.meta.path().is_ident("pre_action")
+}
+
+#[derive(Debug, FromMeta)]
+struct PreActionOptions {
+    exclude: PathList,
+}
+
+pub fn check_for_pre_action(pre_action_info: &mut PreAction, items: &mut Vec<Item>) {
+    let mut num_pre_action_fns = 0;
+    for (_item_index, item) in items.iter_mut().enumerate() {
```

Comment:
What's the point of `enumerate` if the index isn't used?

Reply:
ah, good call. Copied this code from the rest of the macro and modified and didn't realize I didn't end up using it. 👍